File: clase-5/usuarios/backend/src/middleware/auth.middleware.ts
import {NextFunction, Request, Response} from "express";
import jwt from "jsonwebtoken";

interface AuthRequest extends Request {
    user?: any;
}

const authMiddleware = (req: AuthRequest, res: Response, next: NextFunction) => {
    const authHeader = req.headers.authorization;

    if (!authHeader) {
        return res.status(401).json({
            message: "No se proporcionó token de autenticación"
        });
    }

    const token = authHeader.split(" ")[1];

    if (!token) {
        return res.status(401).json({
            message: "Formato de token inválido"
        });
    }

    try {
        req.user = jwt.verify(token, process.env.JWT_SECRET || '');

        next();
    } catch (error) {
        return res.status(401).json({
            message: "Token inválido o expirado"
        });
    }
};

export default authMiddleware;

File: clase-5/empleados/backend/app/middleware/auth_middleware.py
import os
from fastapi import Request, HTTPException, Depends
from fastapi.security import HTTPBearer, HTTPAuthorizationCredentials
from jose import jwt, JWTError
import logging

# Configuración del logger
logging.basicConfig(level=logging.INFO)
logger = logging.getLogger("auth_middleware")

security = HTTPBearer()

class AuthMiddleware:
    async def __call__(self, request: Request, credentials: HTTPAuthorizationCredentials = Depends(security)):
        """
        Middleware para verificar el token JWT en las solicitudes
        """
        if not credentials:
            raise HTTPException(
                status_code=401,
                detail="No se proporcionó token de autenticación"
            )

        token = credentials.credentials

        if not token:
            raise HTTPException(
                status_code=401,
                detail="Formato de token inválido"
            )

        try:
            # Verificar el token
            payload = jwt.decode(
                token,
                os.getenv("JWT_SECRET", "your-secret-key"),
                algorithms=["HS256"]
            )

            # Guardar la información del usuario en el request state
            request.state.user = payload

            return payload
        except JWTError as e:
            logger.error(f"Error al validar token: {str(e)}")
            raise HTTPException(
                status_code=401,
                detail="Token inválido o expirado"
            )

# Instancia del middleware para ser usado como dependencia
auth_middleware = AuthMiddleware()
